add blog for feature LoadBalancerIPMode (#43904)
kubernetes/website: content/en/blog/_posts/2023-12-18-kubernetes-1-29-feature-loadbalancer-ip-mode-alpha.md

This blog introduces `LoadBalancerIPMode`, a new alpha feature in Kubernetes 1.29. 
It provides a configurable approach to define how service implementations, 
exemplified in this blog by kube-proxy, 
handle traffic from pods to the `service.status.loadbalancer.ingress.ip` within the cluster.

## Background

As of today, the different kube-proxy implementations(including ipvs and iptables) bind the external IP 
of the service which type is set to LoadBalancer to each node. 
This is achieved through iptables redirecting packets directly to the service and ipvs binding the IP 
to one interface on the node. This feature is implemented for the following reasons:

1. **Traffic Path Optimization:** Efficiently redirecting pod traffic sent to the load balancer IP 
directly to the backend service by bypassing the load balancer.
  
2. **Handling Load Balancer Packets:** Some load balancers send packets with the destination IP set to 
the load balancer's IP. As a result, these packets need to be routed directly to the correct backend service 
on the node to avoid loops.
  
## Problems

However, there are several problems with the aforementioned behavior:

> 1. **[Source IP Issue]():** Some cloud providers use the load balancer's IP as the source IP when 

Author: The link was missed. I added it in the new commit.

transmitting packets to the node. In the ipvs mode of kube-proxy, 

> there is a problem that health checks from the load balancer never return as the IP is bound to an interface.

Author: The load balancer can reach the listening socket but the reply packets can not out the node. I chane the description here. PTAL.

  
2. **[Feature Loss at Load Balancer Level](https://github.com/kubernetes/kubernetes/issues/66607):** Certain 
cloud providers offer features(such as TLS termination, proxy protocol, etc.) at the load balancer level. 
Bypassing the load balancer results in the loss of these features when the packet reaches the service(leading to protocol errors).
  

Currently, there is a [workaround](https://github.com/kubernetes/kubernetes/issues/66607#issuecomment-474513060) that involves setting the `hostname` of the `service.status.loabalancer.ingress` 
to bypass kube-proxy binding. 
But this is just a makeshift solution.

## Solution

In summary, providing an option for cloud providers to disable the current behavior would be highly beneficial.

> To address this, we propose a solution that introduces a new field `ipMode` of `service.status.loadbalancer.ingress`. 

Reviewer: I made a further suggestion - sorry about the double edit.

---
layout: blog
title: "Kubernetes 1.29: New (alpha) Feature, Load Balancer IP Mode for Services"
date: 2023-12-18
slug: kubernetes-1-29-feature-loadbalancer-ip-mode-alpha
---

**Author:** [Aohan Yang](https://github.com/RyanAoh)

This blog introduces a new alpha feature in Kubernetes 1.29. 
It provides a configurable approach to define how Service implementations, 
exemplified in this blog by kube-proxy, 
handle traffic from pods to the Service, within the cluster.

## Background

> As of today, the different kube-proxy implementations(including ipvs and iptables) bind the external IP 
> of the service which type is set to LoadBalancer to each node. 

Reviewer: ```suggestion
In older Kubernetes releases, the kube-proxy would intercept traffic that was destined for the IP
address associated with a Service of `type: LoadBalancer`. This happened whatever mode you used
for `kube-proxy`.
```

> This is achieved through iptables redirecting packets directly to the service and ipvs binding the IP 

Reviewer: ```suggestion
The interception implemented the expected behavior (traffic eventually reaching the expected
endpoints behind the Service). The mechanism to make that work depended on the mode for kube-prox;
on Linux, kube-proxy in iptables mode would redirecting packets directly to the endpoint; in ipvs mode,
kube-proxy would configure the load balancer's IP address
```
```

> to one interface on the node. This feature is implemented for the following reasons:

Reviewer: ```suggestion
to one interface on the node. The motivation for implementing that interception was for two reasons:
```

> 1. **Traffic path optimization:** Efficiently redirecting pod traffic sent to the load balancer IP 

Reviewer: ```suggestion
1. **Traffic path optimization:** Efficiently redirecting pod traffic - when a container in a pod sends an outbound
   packet that is destined for the load balancer's IP address -
```

> directly to the backend service by bypassing the load balancer.

Reviewer: ```suggestion
  directly to the backend service by bypassing the load balancer.
```

  
2. **Handling load balancer packets:** Some load balancers send packets with the destination IP set to 

> the load balancer's IP. As a result, these packets need to be routed directly to the correct backend service 
> on the node to avoid loops.

Reviewer: ```suggestion
   the load balancer's IP address. As a result, these packets need to be routed directly to the correct backend
   (which might not be local to that node), in order to avoid loops.
```

  
## Problems

However, there are several problems with the aforementioned behavior:

1. **[Source IP](https://github.com/kubernetes/kubernetes/issues/79783):** 
    Some cloud providers use the load balancer's IP as the source IP when 
    transmitting packets to the node. In the ipvs mode of kube-proxy, 
    there is a problem that health checks from the load balancer never return. This occurs because the reply packets 
    would be forward to the local interface `kube-ipvs0`(where the load balancer's IP is bound to) 
    and be subsequently ignored.
  
2. **[Feature loss at load balancer level](https://github.com/kubernetes/kubernetes/issues/66607):**
    Certain cloud providers offer features(such as TLS termination, proxy protocol, etc.) at the
    load balancer level.
    Bypassing the load balancer results in the loss of these features when the packet reaches the service
    (leading to protocol errors).
  

> Currently, there is a [workaround](https://github.com/kubernetes/kubernetes/issues/66607#issuecomment-474513060) that involves setting the `hostname` of the `service.status.loabalancer.ingress` 

Reviewer: ```suggestion
Even with the new alpha behaviour disabled (the default), there is a
[workaround](https://github.com/kubernetes/kubernetes/issues/66607#issuecomment-474513060) that involves setting `.status.loadBalancer.ingress.hostname` for the Service, in order
```
?

to bypass kube-proxy binding. 
But this is just a makeshift solution.

## Solution

In summary, providing an option for cloud providers to disable the current behavior would be highly beneficial.

> To address this, Kubernetes v1.29 introduces a new (alpha) `service.status.loadbalancer.ingress.ipMode` 

Reviewer: ```sugestion
To address this, Kubernetes v1.29 introduces a new (alpha) `.status.loadBalancer.ingress.ipMode` 
```

field for a Service.
This field specifies how the load balancer IP behaves and can be specified only when 
the `.status.loadBalancer.ingress.ip` field is also specified.

Two values are possible for `.status.loadBalancer.ingress.ipMode`: `"VIP"` and `"Proxy"`.
The default value is "VIP", meaning that traffic delivered to the node 
with the destination set to the load balancer's IP and port will be redirected to the backend service by kube-proxy.
This preserves the existing behavior of kube-proxy. 

> The "Proxy" value is intended to prevent kube-proxy from binding the Load Balancer IP to the node in both ipvs and iptables modes. 

Reviewer: ```suggestion
The "Proxy" value is intended to prevent kube-proxy from binding the load balancer's IP address to the node
any mode (iptables, ipvs, nftables).
```
?

What about Windows behavior? There's another operating mode for kube-proxy and the article as written makes it sound like there isn't.

Reviewer: Windows doesn't have nftables, but there is a `kube-proxy` mode for Windows: [`kernelspace`](https://kubernetes.io/docs/reference/networking/virtual-ips/#proxy-mode-kernelspace)

Reviewer: Is there a k/k issue tracked?

Author: I create a new one(https://github.com/kubernetes/kubernetes/issues/122202).

Consequently, traffic is sent directly to the load balancer and then forwarded to the destination node. 
The destination setting for forwarded packets varies depending on how the cloud provider's load balancer delivers traffic:

- If the traffic is delivered to the node then DNATed to the pod, the destination would be set to the node's IP and node port;
- If the traffic is delivered directly to the pod, the destination would be set to the pod's IP and port.

## Usage

Here are the necessary steps to enable this feature:

- Download the [latest Kubernetes project](https://kubernetes.io/releases/download/) (version `v1.29.0` or later).
- Enable the feature gate with the command line flag `--feature-gates=LoadBalancerIPMode=true` 
on kube-proxy, kube-apiserver, and cloud-controller-manager.

> - Set `ipMode` for services which type is set to `LoadBalancer` to the appropriate value. 

Reviewer: ```suggestion
- For Services with `type: LoadBalancer`, set `ipMode` to the appropriate value.
```
?

> This step is likely handled by the cloud-controller-manager during the `EnsureLoadBalancer` process.

Reviewer: ```suggestion
This step is likely handled by your chosen cloud-controller-manager during the `EnsureLoadBalancer` process.
```
(in-tree cloud controller integrations? Not a thing in v1.29)

Author: Yes, it should be user's chosen cloud-controller-manager which may be provided by cloud providers.

## More information

- Read [Specifying IPMode of load balancer status](/docs/concepts/services-networking/service/#load-balancer-ip-mode).

> - Read [KEP-1860 Make Kubernetes aware of the LoadBalancer behaviour](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1860-kube-proxy-IP-node-binding#kep-1860-make-kubernetes-aware-of-the-loadbalancer-behaviour).

Reviewer: ```suggestion
- Read [KEP-1860](https://kep.k8s.io/1860) - [Make Kubernetes aware of the LoadBalancer behaviour](https://github.com/kubernetes/enhancements/tree/b103a6b0992439f996be4314caf3bf7b75652366/keps/sig-network/1860-kube-proxy-IP-node-binding#kep-1860-make-kubernetes-aware-of-the-loadbalancer-behaviour) _(sic)_.
```

## Getting involved

Reach us on [Slack](https://slack.k8s.io/): [#sig-network](https://kubernetes.slack.com/messages/sig-network), 
or through the [mailing list](https://groups.google.com/forum/#!forum/kubernetes-sig-network).

## Acknowledgments

Huge thanks to [@Sh4d1](https://github.com/Sh4d1) for the original KEP and initial implementation code. 
I took over midway and completed the work. Similarly, immense gratitude to other contributors 

> who have assisted in the design, implementation, and review of this feature(alphabetical order):

Reviewer: ```suggestion
who have assisted in the design, implementation, and review of this feature (alphabetical order):
```
